### ISSUE TITLE
Infinite loop in messaging onMessage on Android

### ISSUE THREAD

**Issue by hegelstad** [invertase/react-native-firebase#813]
### Issue

In `Android`, when opening a notification from tray or when app is in foreground, onMessage() is called indefinitely in a loop.

The issue happens when trying to open URLs with react-native's Linking.openURL();

The consecutive following calls have "empty" payloads. Only the first call has the correct payload.

<img width="910" alt="screen shot 2018-02-05 at 15 32 27" src="https://user-images.githubusercontent.com/6997343/35813686-82d30b3e-0a94-11e8-8fe0-7d8dec56abaf.png">

```javascript
apply plugin: "com.android.application"

import com.android.build.OutputFile

/**
 * The react.gradle file registers a task for each build variant (e.g. bundleDebugJsAndAssets
 * and bundleReleaseJsAndAssets).
 * These basically call `react-native bundle` with the correct arguments during the Android build
 * cycle. By default, bundleDebugJsAndAssets is skipped, as in debug/dev mode we prefer to load the
 * bundle directly from the development server. Below you can see all the possible configurations
 * and their defaults. If you decide to add a configuration block, make sure to add it before the
 * `apply from: "../../node_modules/react-native/react.gradle"` line.
 *
 * project.ext.react = [
 *   // the name of the generated asset file containing your JS bundle
 *   bundleAssetName: "index.android.bundle",
 *
 *   // the entry file for bundle generation
 *   entryFile: "index.android.js",
 *
 *   // whether to bundle JS and assets in debug mode
 *   bundleInDebug: false,
 *
 *   // whether to bundle JS and assets in release mode
 *   bundleInRelease: true,
 *
 *   // whether to bundle JS and assets in another build variant (if configured).
 *   // See http://tools.android.com/tech-docs/new-build-system/user-guide#TOC-Build-Variants
 *   // The configuration property can be in the following formats
 *   //         'bundleIn${productFlavor}${buildType}'
 *   //         'bundleIn${buildType}'
 *   // bundleInFreeDebug: true,
 *   // bundleInPaidRelease: true,
 *   // bundleInBeta: true,
 *
 *   // whether to disable dev mode in custom build variants (by default only disabled in release)
 *   // for example: to disable dev mode in the staging build type (if configured)
 *   devDisabledInStaging: true,
 *   // The configuration property can be in the following formats
 *   //         'devDisabledIn${productFlavor}${buildType}'
 *   //         'devDisabledIn${buildType}'
 *
 *   // the root of your project, i.e. where "package.json" lives
 *   root: "../../",
 *
 *   // where to put the JS bundle asset in debug mode
 *   jsBundleDirDebug: "$buildDir/intermediates/assets/debug",
 *
 *   // where to put the JS bundle asset in release mode
 *   jsBundleDirRelease: "$buildDir/intermediates/assets/release",
 *
 *   // where to put drawable resources / React Native assets, e.g. the ones you use via
 *   // require('./image.png')), in debug mode
 *   resourcesDirDebug: "$buildDir/intermediates/res/merged/debug",
 *
 *   // where to put drawable resources / React Native assets, e.g. the ones you use via
 *   // require('./image.png')), in release mode
 *   resourcesDirRelease: "$buildDir/intermediates/res/merged/release",
 *
 *   // by default the gradle tasks are skipped if none of the JS files or assets change; this means
 *   // that we don't look at files in android/ or ios/ to determine whether the tasks are up to
 *   // date; if you have any other folders that you want to ignore for performance reasons (gradle
 *   // indexes the entire tree), add them here. Alternatively, if you have JS files in android/
 *   // for example, you might want to remove it from here.
 *   inputExcludes: ["android/**", "ios/**"],
 *
 *   // override which node gets called and with what additional arguments
 *   nodeExecutableAndArgs: ["node"],
 *
 *   // supply additional arguments to the packager
 *   extraPackagerArgs: []
 * ]
 */

project.ext.react = [
    entryFile: "index.js"
]

apply from: "../../node_modules/react-native/react.gradle"

/**
 * Set this to true to create two separate APKs instead of one:
 *   - An APK that only works on ARM devices
 *   - An APK that only works on x86 devices
 * The advantage is the size of the APK is reduced by about 4MB.
 * Upload all the APKs to the Play Store and people will download
 * the correct one based on the CPU architecture of their device.
 */
def enableSeparateBuildPerCPUArchitecture = false

/**
 * Run Proguard to shrink the Java bytecode in release builds.
 */
def enableProguardInReleaseBuilds = false

android {
    compileSdkVersion 23
    buildToolsVersion "23.0.1"

    defaultConfig {
        applicationId "net.hegelstad.marketplace"
        minSdkVersion 16
        targetSdkVersion 22
        versionCode 1
        versionName "1.0"
        ndk {
            abiFilters "armeabi-v7a", "x86"
        }
    }
    splits {
        abi {
            reset()
            enable enableSeparateBuildPerCPUArchitecture
            universalApk false  // If true, also generate a universal APK
            include "armeabi-v7a", "x86"
        }
    }
    buildTypes {
        release {
            minifyEnabled enableProguardInReleaseBuilds
            proguardFiles getDefaultProguardFile("proguard-android.txt"), "proguard-rules.pro"
        }
    }
    // applicationVariants are e.g. debug, release
    applicationVariants.all { variant ->
        variant.outputs.each { output ->
            // For each separate APK per architecture, set a unique version code as described here:
            // http://tools.android.com/tech-docs/new-build-system/user-guide/apk-splits
            def versionCodes = ["armeabi-v7a":1, "x86":2]
            def abi = output.getFilter(OutputFile.ABI)
            if (abi != null) {  // null for the universal-debug, universal-release variants
                output.versionCodeOverride =
                        versionCodes.get(abi) * 1048576 + defaultConfig.versionCode
            }
        }
    }
}

dependencies {
    compile project(':react-native-vector-icons')
    compile (project(':react-native-firebase')) {
      transitive = false
    }

    compile "com.google.android.gms:play-services-base:11.6.0"
    compile "com.google.firebase:firebase-core:11.6.0"
    compile "com.google.firebase:firebase-auth:11.6.0"
    compile "com.google.firebase:firebase-firestore:11.6.0"
    compile "com.google.firebase:firebase-messaging:11.6.0"

    compile fileTree(dir: "libs", include: ["*.jar"])
    compile "com.android.support:appcompat-v7:23.0.1"
    compile "com.facebook.react:react-native:+"  // From node_modules
    compile 'me.leolin:ShortcutBadger:1.1.18@aar'
}

// Run this once to be able to run the application with BUCK
// puts all compile dependencies into folder libs for BUCK to use
task copyDownloadableDepsToLibs(type: Copy) {
    from configurations.compile
    into 'libs'
}

apply plugin: 'com.google.gms.google-services'

```

### Environment

Android 

1. Application Target Platform: macOS Sierra

2. Development Operating System: `react-native run-android`, genymotion, samsung S6 API 23.

3. Build Tools: ?

4. React Native version: `^0.52.1`

5. RNFirebase Version: `^3.2.2`

6. Firebase Module: auth, firestore, messaging, cloud functions

**Comment by chrisbianca** [comment]
@hegelstad can we see your code usage?

**Comment by hegelstad** [comment]
```javascript
    firebase.messaging().onMessage(payload => {
      console.log('Opened when app is alive');
      console.log(payload);
     
      const prefix = Platform.OS == 'android' ? 'ifimarketplace://ifimarketplace/' : 'ifimarketplace://'
      const url = `${prefix}tabs/messages/${payload.key}`;
      console.log(url);
      Linking.openURL(url).catch(err => console.error(err));
});
```

**Comment by chrisbianca** [comment]
@hegelstad my guess is that `onMessage` is picking up the `Linking.openUrl()` result, as I assume this is opening something within your app?

We're currently in the process of rewriting the internals of the messaging library as there are lots of small issues like this one.  You can monitor progress here: https://github.com/invertase/react-native-firebase/issues/595

At the moment the intent that loads `onMessage` isn't scoped to just get message events, something that we're fixing as part of this work.

**Comment by hegelstad** [comment]
@chrisbianca That sounds about right. I am using it for deep linking functionality with ReactNavigation.

Do you have any idea of a fix that I could try to use? And what is the time frame for this fix? I need this functionality for my thesis work before the end of february.. 

Is [https://rnfirebase.io/docs/v3.2.x/links/reference/links](https://rnfirebase.io/docs/v3.2.x/links/reference/links) the same as deep linking?

**Comment by chrisbianca** [comment]
I don't have an exact date, but it should be sometime this month if all goes well.

Yes, https://rnfirebase.io/docs/v3.2.x/links/reference/links is very similar to react-native's deep linking

**Comment by hegelstad** [comment]
Okay, thanks, I will see if I can get react-native-firebase links to work with react-navigation. Appreciate your swift response.

**Comment by chrisbianca** [comment]
Closing this issue as it will be resolved as part of the work on #595

**Comment by chrisbianca** [comment]
Good news, the long awaited alpha of our messaging and notifications overhaul is now available!!

Check out the release notes here: https://github.com/invertase/react-native-firebase/releases/tag/v4.0.0-alpha.1

If you have any comments and suggestions or want to report an issue, come find us on [Discord](https://discord.gg/C9aK28N)
